src/callbacks/memory_profiler.cpp

```diff
@@ -0,0 +1,578 @@
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+//
+// memory_profiler .hpp .cpp - Itemized memory usage profiling.
+///////////////////////////////////////////////////////////////////////////////
+
+#include "lbann/callbacks/memory_profiler.hpp"
+#include "lbann/layers/data_type_layer.hpp"
+#include "lbann/models/model.hpp"
+#include "lbann/utils/gpu/helpers.hpp"
+#include "lbann/utils/serialize.hpp"
+#include "lbann/weights/data_type_weights.hpp"
+#include "lbann/weights/weights.hpp"
+
+#include "lbann/proto/callbacks.pb.h"
+
+#include <algorithm>
+#include <string>
+
+namespace lbann {
+namespace callback {
+
+/**
+ * @brief Returns the currently used memory, or 0 if LBANN was not compiled with
+ * GPU support.
+ */
+static inline size_t get_used_gpu_memory()
+{
+#ifdef LBANN_HAS_GPU
+  size_t available;
+  size_t total;
+#ifdef LBANN_HAS_CUDA
+  FORCE_CHECK_CUDA(cudaMemGetInfo(&available, &total));
+#elif defined(LBANN_HAS_ROCM)
+  FORCE_CHECK_ROCM(hipMemGetInfo(&available, &total));
+#endif
+  // TODO(later): Might be nicer to return a struct with gathered information
+  // (min, max, median across ranks)
+  return total - available;
+#else
+  return 0;
+#endif
+}
+
+/**
+ * @brief Returns the total memory, or 0 if LBANN was not compiled with
+ * GPU support.
+ */
+static inline size_t get_total_gpu_memory()
+{
+#ifdef LBANN_HAS_GPU
+  size_t available;
+  size_t total;
+#ifdef LBANN_HAS_CUDA
+  FORCE_CHECK_CUDA(cudaMemGetInfo(&available, &total));
+#elif defined(LBANN_HAS_ROCM)
+  FORCE_CHECK_ROCM(hipMemGetInfo(&available, &total));
+#endif
+  return total;
+#else
+  return 0;
+#endif
+}
+
+/**
+ * @brief Prints out the shape and allocated size of a matrix to the stream
+ * given in the second argument. Returns the allocated size as well.
+ */
+template <typename T>
+static inline size_t report_dist_matrix(const El::AbstractDistMatrix<T>& m,
+                                        std::stringstream& stream)
+{
+  size_t allocated = m.AllocatedMemory() * sizeof(T);
+  stream << m.Height() << " x " << m.Width()
+         << " (local shape: " << m.LocalHeight() << " x " << m.LocalWidth()
+         << "). Size: " << allocated / 1048576.0 << " MiB" << std::endl;
+  return allocated;
+}
+
+memory_profiler::memory_profiler(bool detailed_first_step)
+  : callback_base(), m_detailed_first_step(detailed_first_step)
+{
+#ifndef LBANN_HAS_GPU
+  LBANN_WARNING("Memory profiler callback only provides unaccounted memory "
+                "information with GPU support.");
```

Comment:
I am having trouble parsing this warning text. Maybe rephrase?

Reply:
Fixed, thanks.